p2p/src/error.rs

```diff
@@ -14,124 +14,133 @@
 // limitations under the License.
 //
 // Author(s): A. Altonen
+use libp2p::{
+    gossipsub::error::PublishError as GossipsubPublishError,
+    swarm::{handler::ConnectionHandlerUpgrErr, DialError::*},
+};
 use thiserror::Error;
 
-// TODO: think about which errors should be returned and when
-// TODO: store peerid where appropriate!
 #[derive(Error, Debug, PartialEq, Eq)]
 pub enum ProtocolError {
+    #[error("Peer is in different network")]
     DifferentNetwork,
+    #[error("Peer has an unsupported version")]
     InvalidVersion,
+    #[error("Peer sent an invalid message")]
     InvalidMessage,
+    #[error("Peer is incompatible")] // TODO: remove?
     Incompatible,
+    #[error("Peer is unresponsive")]
     Unresponsive,
+    #[error("Peer uses an invalid protocol")] // TODO: remove?
     InvalidProtocol,
+    #[error("Peer is an unknown network")] // TODO: remove?
     UnknownNetwork,
+    #[error("Peer is in an invalid state to perform this operation")]
     InvalidState,
 }
 
-// TODO: refactor error code
 #[derive(Error, Debug, PartialEq, Eq)]
-pub enum Libp2pError {
-    #[error("NoiseError: `{0:?}`")]
-    NoiseError(String),
-    #[error("TransportError: `{0:?}`")]
-    TransportError(String),
-    #[error("DialError: `{0:?}`")]
-    DialError(String),
-    #[error("SubscriptionError: `{0:?}`")]
-    SubscriptionError(String),
-    #[error("PublishError: `{0:?}`")]
-    PublishError(String),
-    #[error("IdentifyError: `{0:?}`")]
-    IdentifyError(String),
+pub enum PeerError {
+    #[error("Peer disconnected")]
+    PeerDisconnected,
+    #[error("No peers")]
+    NoPeers,
+    #[error("Peer doesn't exist")]
+    PeerDoesntExist,
+    #[error("Peer already exists")]
+    PeerExists,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum PublishError {
+    #[error("Message has already been published")]
+    Duplicate,
+    #[error("Failed to sign message")]
+    SigningFailed,
+    #[error("Not enough peers in topic")]
+    InsufficientPeers,
+    #[error("Message is too large")]
+    MessageTooLarge,
+    #[error("Failed to compress the message")]
+    TransformFailed,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum DialError {
+    #[error("Peer is banned")]
+    Banned,
+    #[error("Limit for outgoing connections reached")]
+    ConnectionLimit,
+    #[error("Tried to dial local node")]
+    LocalPeerId,
```

Comment:
This doesn't sound like a crime unless one is call themselves. Functional tests do call local nodes. Why is this an error? Did you mean `AttemptedToDialToSelf`?

Reply:
In libp2p, local refers to self so if the dial target is local peer ID, it means that the we're trying to dial "self". But I can change our error to say self